### ISSUE TITLE
FAQ/Design: stateful transform for exog across datasets

### ISSUE THREAD

**Issue by josef-pkt** [statsmodels/statsmodels#5347]
When we use statefull transformation in the creation of the design matrix, `exog`, then the parameterization is data dependent.

This is the case with categorical variables if not all levels are available in all datasets. However, in this case the parameters for the missing levels cannot be estimated.

Another example are splines with data dependent knot location.

A polynomial basis will be statefull if it includes data dependent transformation. In R polynomials in formulas are orthogonalized and the orthogonalizing transform is statefull. When using numpy polynomial it is possible to map the data into a fixed domain for better numerical properties. Because this mapping depends on the data, this transformation is also stateful.
In contrast simple polynomial terms like x, x**2, x**3, ... are not stateful but have widely varying values if x is not restricted to small numbers.

When we want to have the same parameterization across datasets, for example to compare the estimated parameters across datasets, then we need to fix the parameterization of the stateful transformations before creating individual models.
For example, we can choose the knots based on the entire set of data and use the same knots when creating and estimating individual models.
If we are only interested in the prediction, then the choosing a common parameterization is less important and we can choose an appropriate model separately for each dataset.

**Comment by josef-pkt** [comment]
Adding stateful transform classes to statsmodels?
Currently we have only the version in the GAM PR.

scikit-learn also has a set of classes for it that are stateful. It uses a fit function that can also optimize to estimate the parameters for the transformation, e.g. for Box-Cox
https://github.com/scikit-learn/scikit-learn/blob/55bf5d9/sklearn/preprocessing/data.py#L2482
https://stackoverflow.com/questions/53624804/how-to-normalize-a-non-normal-distribution
https://groups.google.com/d/msg/pystatsmodels/BINdYwHHSXg/y0hnb-ahBgAJ

patsy uses `memorize_finish` instead of `fit`

In my GAM version, there is currently no separate method to call. Initializing the parameters is done in `__init__` immediately, which might not be a good idea. (E.g. open question when to include a constant in the spline basis. Currently, it cannot be chosen in the context of the specific model and is predefined by user or hardcoded in the classes.
